src/fb.c

```diff
@@ -1,63 +1,275 @@
+        return;
+
+    console.text[fb_console_get_row(console.cursor.row)][console.cursor.col] = c;
+    fb_console_blit_char(console.cursor);
+
+    console.cursor.col++;
+}
+
+static void fb_check_scroll(void)
+{
+    if (console.cursor.row < console.n_rows)
+        return;
+
+    memset(console.text[fb_console_get_row(0)], 0, CONSOLE_MAX_COLS);
+    console.row_offset = (console.row_offset + 1) % console.n_rows;
+    console.cursor.row--;
+
+    assert(console.cursor.row < console.n_rows);
+
+    fb_console_blit_all();
+}
+
+void fb_console_putc(u8 c)
+{
+    if (!smp_is_primary())
+        return;
+
+    if (c == '\n') {
+        console.cursor.row++;
+        console.cursor.col = 0;
+    } else if (c == '\r') {
+        console.cursor.col = 0;
+    } else if (c >= 0x20 && c <= 0x7e) {
+        fb_console_putbyte(c);
+    } else {
+        fb_console_putbyte('?');
+    }
+
+    fb_check_scroll();
+}
+
+void fb_console_write(const void *buf, size_t count)
+{
+    const u8 *p = buf;
+
+    while (count--)
+        fb_console_putc(*p++);
+}
+
+void fb_console_init(void)
+{
+    if (cur_boot_args.video.depth & FB_DEPTH_FLAG_RETINA) {
+        console.font.ptr = _binary_build_font_retina_bin_start;
+        console.font.width = 16;
+        console.font.height = 32;
+    } else {
+        console.font.ptr = _binary_build_font_bin_start;
+        console.font.width = 8;
+        console.font.height = 16;
+    }
+
+    console.n_rows =
+        min(CONSOLE_MAX_ROWS,
+            (fb.height - CONSOLE_MARGIN_TOP - CONSOLE_MARGIN_BOTTOM) / console.font.height);
+    console.n_cols =
+        min(CONSOLE_MAX_COLS, (fb.width - CONSOLE_MARGIN_LEFT - CONSOLE_MARGIN_RIGHT - logo.width) /
+                                  console.font.width);
+
+    if (console.row_offset > 0) {
+        /*
+         * TODO: handle this case better if we ever print more than CONSOLE_MAX_ROWS before calling
+         * fb_console_init
+         */
+        printf("fb console: printed more than %d rows before init, discarding early console.\n",
+               CONSOLE_MAX_ROWS);
+        memset(console.text, 0, sizeof(console.text));
+        console.cursor.col = 0;
+        console.cursor.row = 0;
+    }
+
+    /*
+     * this size of our console may have shrunk and we'll need to throw away some early rows
+     * and copy later ones to the start.
+     * we can't use fb_check_scroll() here because we will need to copy rows that are already
+     * larger than console.n_rows.
+     */
+    if (console.cursor.row > console.n_rows) {
+        u32 row_offset = console.cursor.row - console.n_rows;
+        for (u32 row = 0; row < console.n_rows; ++row)
+            memcpy(console.text[row], console.text[row + row_offset], CONSOLE_MAX_COLS);
+        console.cursor.row = console.n_rows;
+    }
+    fb_check_scroll();
+
+    console.initialized = 1;
+    fb_console_blit_all();
+}
+
+void fb_blit(u32 x, u32 y, u32 w, u32 h, void *data, u32 stride)
```

Comment:
The stride here is a pixel stride for the source image right? I think that parameter should be before the `void* data` parameter. 
Also it's not really a `void*` since you only ever use it as a `u8*`, and all the things you pass in are also defined that way so I think it's best to not have it be a void*

Reply:
same argument like making fb static: probably a good idea but feels like it's out of scope for this change.

Reply:
I don't really see how it's out of scope, this function's parameters were changed in this series of commits, that's where the stride parameter was introduced.

Reply:
No, the stride parameter has always been there. the only change was `int` -> `u32`.

Reply:
You're right I'm not sure how I missed that. My bad sorry about that.

Reply:
No worries, it's probably github's fault. The diff for all commits together looks very confusing while the diff for the individual commit looks fine.